File: app/fsharp/src/Ch11/TravelGuide.fs
namespace Ch11

/// 実践的なアプリケーション構築
/// Scala の TravelGuide に相当する機能を F# で実現
module TravelGuide =

    open System

    // ============================================
    // ドメインモデル
    // ============================================

    /// ロケーション ID（値オブジェクト）
    type LocationId = LocationId of string

    module LocationId =
        let value (LocationId id) = id
        let create id = LocationId id

    /// ロケーション
    type Location = {
        Id: LocationId
        Name: string
        Population: int
    }

    /// アトラクション（観光地）
    type Attraction = {
        Name: string
        Description: string option
        Location: Location
    }

    /// ミュージックアーティスト
    type MusicArtist = {
        Name: string
        Genre: string option
    }

    /// 映画
    type Movie = {
        Name: string
        BoxOffice: int option
    }

    /// 旅行ガイド
    type TravelGuide = {
        Attraction: Attraction
        Subjects: string list
    }

    /// アトラクションのソート順
    type AttractionOrdering =
        | ByName
        | ByLocationPopulation

    // ============================================
    // データアクセス層の抽象化
    // ============================================

    /// データアクセスインターフェース
    type IDataAccess =
        abstract member FindAttractions: name: string * ordering: AttractionOrdering * limit: int -> Async<Attraction list>
        abstract member FindArtistsFromLocation: locationId: LocationId * limit: int -> Async<MusicArtist list>
        abstract member FindMoviesAboutLocation: locationId: LocationId * limit: int -> Async<Movie list>

    // ============================================
    // リソース管理（IDisposable パターン）
    // ============================================

    /// リソースを安全に管理するための型
    type Resource<'a> = {
        Acquire: unit -> Async<'a>
        Release: 'a -> Async<unit>
    }

    module Resource =
        /// リソースを作成
        let make (acquire: unit -> Async<'a>) (release: 'a -> Async<unit>) : Resource<'a> =
            { Acquire = acquire; Release = release }

        /// リソースを使用して操作を実行
        let useAsync (f: 'a -> Async<'b>) (resource: Resource<'a>) : Async<'b> =
            async {
                let! r = resource.Acquire()
                try
                    return! f r
                finally
                    resource.Release r |> Async.RunSynchronously
            }

        /// 2つのリソースを結合
        let bind (f: 'a -> Resource<'b>) (resource: Resource<'a>) : Resource<'b> =
            {
                Acquire = fun () ->
                    async {
                        let! a = resource.Acquire()
                        let resourceB = f a
                        return! resourceB.Acquire()
                    }
                Release = fun b ->
                    async {
                        // Note: In a real implementation, we'd need to track both resources
                        return ()
                    }
            }

    // ============================================
    // キャッシュの実装
    // ============================================

    open Ch10.ConcurrentProcessing

    /// キャッシュ付きデータアクセス
    let cachedDataAccess (dataAccess: IDataAccess) : Async<IDataAccess> =
        async {
            let attractionCache = Ref.Of(Map.empty<string, Attraction list>)
            let artistCache = Ref.Of(Map.empty<string, MusicArtist list>)
            let movieCache = Ref.Of(Map.empty<string, Movie list>)

            return
                { new IDataAccess with
                    member _.FindAttractions(name, ordering, limit) =
                        async {
                            let key = sprintf "%s-%A-%d" name ordering limit
                            let cached = attractionCache.Get() |> Map.tryFind key
                            match cached with
                            | Some attractions -> return attractions
                            | None ->
                                let! attractions = dataAccess.FindAttractions(name, ordering, limit)
                                attractionCache.Update(Map.add key attractions)
                                return attractions
                        }

                    member _.FindArtistsFromLocation(locationId, limit) =
                        async {
                            let key = sprintf "%s-%d" (LocationId.value locationId) limit
                            let cached = artistCache.Get() |> Map.tryFind key
                            match cached with
                            | Some artists -> return artists
                            | None ->
                                let! artists = dataAccess.FindArtistsFromLocation(locationId, limit)
                                artistCache.Update(Map.add key artists)
                                return artists
                        }

                    member _.FindMoviesAboutLocation(locationId, limit) =
                        async {
                            let key = sprintf "%s-%d" (LocationId.value locationId) limit
                            let cached = movieCache.Get() |> Map.tryFind key
                            match cached with
                            | Some movies -> return movies
                            | None ->
                                let! movies = dataAccess.FindMoviesAboutLocation(locationId, limit)
                                movieCache.Update(Map.add key movies)
                                return movies
                        }
                }
        }

    // ============================================
    // アプリケーションロジック
    // ============================================

    /// 旅行ガイドを生成
    let travelGuide (data: IDataAccess) (attractionName: string) : Async<TravelGuide option> =
        async {
            let! attractions = data.FindAttractions(attractionName, ByLocationPopulation, 1)
            match attractions with
            | [] -> return None
            | attraction :: _ ->
                let! artists = data.FindArtistsFromLocation(attraction.Location.Id, 2)
                let! movies = data.FindMoviesAboutLocation(attraction.Location.Id, 2)
                let subjects =
                    (artists |> List.map (fun a -> a.Name))
                    @ (movies |> List.map (fun m -> m.Name))
                return Some {
                    Attraction = attraction
                    Subjects = subjects
                }
        }

    /// 複数のアトラクションから旅行ガイドを生成
    let travelGuides (data: IDataAccess) (attractionName: string) (limit: int) : Async<TravelGuide list> =
        async {
            let! attractions = data.FindAttractions(attractionName, ByLocationPopulation, limit)
            let! guides =
                attractions
                |> List.map (fun attraction ->
                    async {
                        let! artists = data.FindArtistsFromLocation(attraction.Location.Id, 2)
                        let! movies = data.FindMoviesAboutLocation(attraction.Location.Id, 2)
                        let subjects =
                            (artists |> List.map (fun a -> a.Name))
                            @ (movies |> List.map (fun m -> m.Name))
                        return {
                            Attraction = attraction
                            Subjects = subjects
                        }
                    })
                |> Async.Parallel
            return guides |> Array.toList
        }

    // ============================================
    // テスト用のスタブ実装
    // ============================================

    /// テスト用のデータアクセス実装
    let testDataAccess (testAttractions: Attraction list) (testArtists: MusicArtist list) (testMovies: Movie list) : IDataAccess =
        { new IDataAccess with
            member _.FindAttractions(name, _, limit) =
                async {
                    return
                        testAttractions
                        |> List.filter (fun a -> a.Name.Contains(name))
                        |> List.truncate limit
                }

            member _.FindArtistsFromLocation(_, limit) =
                async { return testArtists |> List.truncate limit }

            member _.FindMoviesAboutLocation(_, limit) =
                async { return testMovies |> List.truncate limit }
        }

    // ============================================
    // サンプルデータ
    // ============================================

    let sampleLocation = {
        Id = LocationId "Q123"
        Name = "Tokyo"
        Population = 14000000
    }

    let sampleAttraction = {
        Name = "Tokyo Tower"
        Description = Some "Famous landmark in Tokyo"
        Location = sampleLocation
    }

    let sampleArtist = {
        Name = "Hikaru Utada"
        Genre = Some "Pop"
    }

    let sampleMovie = {
        Name = "Lost in Translation"
        BoxOffice = Some 44000000
    }

File: app/csharp/src/Ch11/TravelGuide.cs
using LanguageExt;
using static LanguageExt.Prelude;
using Ch10;

namespace Ch11;

/// <summary>
/// 第11章: 実践的なアプリケーション構築
/// TravelGuide アプリケーションの実装
/// </summary>
public static class TravelGuideApp
{
    // ============================================
    // ドメインモデル
    // ============================================

    /// <summary>
    /// ロケーション ID（値オブジェクト）
    /// </summary>
    public readonly record struct LocationId(string Value)
    {
        public static LocationId Create(string id) => new(id);
        public override string ToString() => Value;
    }

    /// <summary>
    /// ロケーション
    /// </summary>
    public record Location(LocationId Id, string Name, int Population);

    /// <summary>
    /// アトラクション（観光地）
    /// </summary>
    public record Attraction(string Name, Option<string> Description, Location Location);

    /// <summary>
    /// ミュージックアーティスト
    /// </summary>
    public record MusicArtist(string Name, Option<string> Genre);

    /// <summary>
    /// 映画
    /// </summary>
    public record Movie(string Name, Option<int> BoxOffice);

    /// <summary>
    /// 旅行ガイド
    /// </summary>
    public record TravelGuide(Attraction Attraction, Seq<string> Subjects);

    /// <summary>
    /// アトラクションのソート順
    /// </summary>
    public enum AttractionOrdering
    {
        ByName,
        ByLocationPopulation
    }

    // ============================================
    // データアクセス層の抽象化
    // ============================================

    /// <summary>
    /// データアクセスインターフェース
    /// </summary>
    public interface IDataAccess
    {
        Task<Seq<Attraction>> FindAttractions(string name, AttractionOrdering ordering, int limit);
        Task<Seq<MusicArtist>> FindArtistsFromLocation(LocationId locationId, int limit);
        Task<Seq<Movie>> FindMoviesAboutLocation(LocationId locationId, int limit);
    }

    // ============================================
    // リソース管理
    // ============================================

    /// <summary>
    /// リソースを安全に管理するための型
    /// </summary>
    public class Resource<T>
    {
        private readonly Func<Task<T>> _acquire;
        private readonly Func<T, Task> _release;

        private Resource(Func<Task<T>> acquire, Func<T, Task> release)
        {
            _acquire = acquire;
            _release = release;
        }

        /// <summary>
        /// リソースを作成
        /// </summary>
        public static Resource<T> Make(Func<Task<T>> acquire, Func<T, Task> release)
            => new(acquire, release);

        /// <summary>
        /// リソースを使用して操作を実行
        /// </summary>
        public async Task<TResult> UseAsync<TResult>(Func<T, Task<TResult>> f)
        {
            var resource = await _acquire();
            try
            {
                return await f(resource);
            }
            finally
            {
                await _release(resource);
            }
        }

        /// <summary>
        /// リソースを使用して操作を実行（同期版）
        /// </summary>
        public async Task<TResult> Use<TResult>(Func<T, TResult> f)
        {
            var resource = await _acquire();
            try
            {
                return f(resource);
            }
            finally
            {
                await _release(resource);
            }
        }
    }

    /// <summary>
    /// Resource 用の静的ヘルパー
    /// </summary>
    public static class Resource
    {
        public static Resource<T> Make<T>(Func<Task<T>> acquire, Func<T, Task> release)
            => Resource<T>.Make(acquire, release);
    }

    // ============================================
    // キャッシュの実装
    // ============================================

    /// <summary>
    /// キャッシュ付きデータアクセスを作成
    /// </summary>
    public static IDataAccess CachedDataAccess(IDataAccess dataAccess)
    {
        var attractionCache = ConcurrentProcessing.Ref<Map<string, Seq<Attraction>>>.Of(Map<string, Seq<Attraction>>());
        var artistCache = ConcurrentProcessing.Ref<Map<string, Seq<MusicArtist>>>.Of(Map<string, Seq<MusicArtist>>());
        var movieCache = ConcurrentProcessing.Ref<Map<string, Seq<Movie>>>.Of(Map<string, Seq<Movie>>());

        return new CachedDataAccessImpl(dataAccess, attractionCache, artistCache, movieCache);
    }

    private class CachedDataAccessImpl : IDataAccess
    {
        private readonly IDataAccess _dataAccess;
        private readonly ConcurrentProcessing.Ref<Map<string, Seq<Attraction>>> _attractionCache;
        private readonly ConcurrentProcessing.Ref<Map<string, Seq<MusicArtist>>> _artistCache;
        private readonly ConcurrentProcessing.Ref<Map<string, Seq<Movie>>> _movieCache;

        public CachedDataAccessImpl(
            IDataAccess dataAccess,
            ConcurrentProcessing.Ref<Map<string, Seq<Attraction>>> attractionCache,
            ConcurrentProcessing.Ref<Map<string, Seq<MusicArtist>>> artistCache,
            ConcurrentProcessing.Ref<Map<string, Seq<Movie>>> movieCache)
        {
            _dataAccess = dataAccess;
            _attractionCache = attractionCache;
            _artistCache = artistCache;
            _movieCache = movieCache;
        }

        public async Task<Seq<Attraction>> FindAttractions(string name, AttractionOrdering ordering, int limit)
        {
            var key = $"{name}-{ordering}-{limit}";
            var cached = _attractionCache.Get().Find(key);
            if (cached.IsSome)
            {
                return cached.IfNone(Seq<Attraction>());
            }

            var attractions = await _dataAccess.FindAttractions(name, ordering, limit);
            _attractionCache.Update(cache => cache.Add(key, attractions));
            return attractions;
        }

        public async Task<Seq<MusicArtist>> FindArtistsFromLocation(LocationId locationId, int limit)
        {
            var key = $"{locationId.Value}-{limit}";
            var cached = _artistCache.Get().Find(key);
            if (cached.IsSome)
            {
                return cached.IfNone(Seq<MusicArtist>());
            }

            var artists = await _dataAccess.FindArtistsFromLocation(locationId, limit);
            _artistCache.Update(cache => cache.Add(key, artists));
            return artists;
        }

        public async Task<Seq<Movie>> FindMoviesAboutLocation(LocationId locationId, int limit)
        {
            var key = $"{locationId.Value}-{limit}";
            var cached = _movieCache.Get().Find(key);
            if (cached.IsSome)
            {
                return cached.IfNone(Seq<Movie>());
            }

            var movies = await _dataAccess.FindMoviesAboutLocation(locationId, limit);
            _movieCache.Update(cache => cache.Add(key, movies));
            return movies;
        }
    }

    // ============================================
    // アプリケーションロジック
    // ============================================

    /// <summary>
    /// 旅行ガイドを生成
    /// </summary>
    public static async Task<Option<TravelGuide>> GetTravelGuide(IDataAccess data, string attractionName)
    {
        var attractions = await data.FindAttractions(attractionName, AttractionOrdering.ByLocationPopulation, 1);
        if (attractions.IsEmpty)
        {
            return None;
        }

        var attraction = attractions.First();
        var artists = await data.FindArtistsFromLocation(attraction.Location.Id, 2);
        var movies = await data.FindMoviesAboutLocation(attraction.Location.Id, 2);

        var subjects = artists.Map(a => a.Name).Concat(movies.Map(m => m.Name));

        return Some(new TravelGuide(attraction, subjects));
    }

    /// <summary>
    /// 複数のアトラクションから旅行ガイドを生成
    /// </summary>
    public static async Task<Seq<TravelGuide>> GetTravelGuides(IDataAccess data, string attractionName, int limit)
    {
        var attractions = await data.FindAttractions(attractionName, AttractionOrdering.ByLocationPopulation, limit);

        var guideTasks = attractions.Map(async attraction =>
        {
            var artists = await data.FindArtistsFromLocation(attraction.Location.Id, 2);
            var movies = await data.FindMoviesAboutLocation(attraction.Location.Id, 2);
            var subjects = artists.Map(a => a.Name).Concat(movies.Map(m => m.Name));
            return new TravelGuide(attraction, subjects);
        });

        var guides = await Task.WhenAll(guideTasks);
        return toSeq(guides);
    }

    // ============================================
    // テスト用のスタブ実装
    // ============================================

    /// <summary>
    /// テスト用のデータアクセス実装
    /// </summary>
    public static IDataAccess TestDataAccess(
        Seq<Attraction> testAttractions,
        Seq<MusicArtist> testArtists,
        Seq<Movie> testMovies)
    {
        return new TestDataAccessImpl(testAttractions, testArtists, testMovies);
    }

    private class TestDataAccessImpl : IDataAccess
    {
        private readonly Seq<Attraction> _attractions;
        private readonly Seq<MusicArtist> _artists;
        private readonly Seq<Movie> _movies;

        public TestDataAccessImpl(Seq<Attraction> attractions, Seq<MusicArtist> artists, Seq<Movie> movies)
        {
            _attractions = attractions;
            _artists = artists;
            _movies = movies;
        }

        public Task<Seq<Attraction>> FindAttractions(string name, AttractionOrdering ordering, int limit)
        {
            var filtered = _attractions
                .Filter(a => a.Name.Contains(name))
                .Take(limit);
            return Task.FromResult(filtered);
        }

        public Task<Seq<MusicArtist>> FindArtistsFromLocation(LocationId locationId, int limit)
        {
            return Task.FromResult(_artists.Take(limit));
        }

        public Task<Seq<Movie>> FindMoviesAboutLocation(LocationId locationId, int limit)
        {
            return Task.FromResult(_movies.Take(limit));
        }
    }

    // ============================================
    // サンプルデータ
    // ============================================

    public static Location SampleLocation => new(
        LocationId.Create("Q123"),
        "Tokyo",
        14000000
    );

    public static Attraction SampleAttraction => new(
        "Tokyo Tower",
        Some("Famous landmark in Tokyo"),
        SampleLocation
    );

    public static MusicArtist SampleArtist => new(
        "Hikaru Utada",
        Some("Pop")
    );

    public static Movie SampleMovie => new(
        "Lost in Translation",
        Some(44000000)
    );
}
